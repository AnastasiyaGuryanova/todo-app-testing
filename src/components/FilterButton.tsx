import FilterIcon from "../icons/filter.png";

type Props = {
  onClick: () => void;
  disabled: boolean;
};

export const FilterButton = ({ onClick, disabled }: Props) => {
  return (
    <button
      className="button button-with-icon"
      onClick={onClick}
      disabled={disabled}
      data-alt="фильтровать задачи"
    >
      <img src={FilterIcon} alt="Фильтр" />
    </button>
  );
};
