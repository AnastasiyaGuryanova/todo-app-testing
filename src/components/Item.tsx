import { DeleteButton } from "./DeleteButton";

type Props = Task & {
  onDelete: (id: Task["id"]) => void;
  onToggle: (id: Task["id"]) => void;
};

export const Item = (props: Props) => {
  const effectiveHeader =
    props.header.trim() === "" ? "Без названия" : props.header;
  const truncatedHeader =
    effectiveHeader.length > 32
      ? effectiveHeader.slice(0, 32)
      : effectiveHeader;

  return (
    <li className="item-wrapper">
      <input
        type="checkbox"
        id={props.id}
        defaultChecked={props.done}
        onChange={() => props.onToggle(props.id)}
      />

      <label htmlFor={props.id} onClick={() => props.onToggle(props.id)}>
        {props.done ? <s>{truncatedHeader}</s> : truncatedHeader}
      </label>

      <DeleteButton
        disabled={!props.done}
        onClick={() => props.onDelete(props.id)}
      />
    </li>
  );
};
