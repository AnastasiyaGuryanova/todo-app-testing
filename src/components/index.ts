export * from "./AddButton";
export * from "./DeleteButton";
export * from "./Empty";
export * from "./FilterButton";
export * from "./Input";
export * from "./Item";
export * from "./List";
export * from "./Notifier";
