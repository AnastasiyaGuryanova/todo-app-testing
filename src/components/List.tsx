import { Item } from "./Item";

type Props = {
  items: Task[];
  onDelete: (id: Task["id"]) => void;
  onToggle: (id: Task["id"]) => void;
};

export const List = ({ items, onDelete, onToggle }: Props) => {
  const visibleItems = items.filter((item) => !item.done).slice(0, 10);

  return (
    <ul className="task-list tasks">
      {visibleItems.map((item) => (
        <Item {...item} key={item.id} onDelete={onDelete} onToggle={onToggle} />
      ))}
    </ul>
  );
};
