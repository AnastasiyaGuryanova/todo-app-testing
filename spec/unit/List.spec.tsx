import { render, screen, fireEvent } from "@testing-library/react";
import { List } from "src/components/List";

describe("List", () => {
  it("отображение списка задач", () => {
    const onDelete = jest.fn();
    const onToggle = jest.fn();

    const items: Task[] = [
      {
        id: "1",
        header: "купить хлеб",
        done: false,
      },
      {
        id: "2",
        header: "купить молоко",
        done: false,
      },
      {
        id: "3",
        header: "выгулять собаку",
        done: true,
      },
    ];

    const { rerender, asFragment } = render(
      <List items={items} onDelete={onDelete} onToggle={onToggle} />
    );
    const firstRender = asFragment();

    expect(screen.getByText("купить хлеб")).toBeInTheDocument();
    expect(screen.getByText("купить молоко")).toBeInTheDocument();
    expect(screen.getByText("выгулять собаку")).toBeInTheDocument();

    const updatedItems = items.slice(0, 2);

    rerender(
      <List items={updatedItems} onDelete={onDelete} onToggle={onToggle} />
    );
    const secondRender = asFragment();

    expect(screen.getByText("купить хлеб")).toBeInTheDocument();
    expect(screen.getByText("купить молоко")).toBeInTheDocument();
    expect(screen.queryByText("выгулять собаку")).not.toBeInTheDocument();

    expect(firstRender).toMatchDiffSnapshot(secondRender);
  });

  it("корректно вызывает обработчики при взаимодействии с задачами", () => {
    const onDelete = jest.fn();
    const onToggle = jest.fn();

    const items: Task[] = [
      {
        id: "1",
        header: "купить хлеб",
        done: false,
      },
      {
        id: "2",
        header: "купить молоко",
        done: true,
      },
    ];

    render(<List items={items} onDelete={onDelete} onToggle={onToggle} />);

    const firstCheckbox = screen.getAllByRole("checkbox")[0];
    fireEvent.click(firstCheckbox);
    expect(onToggle).toHaveBeenCalledWith("1");

    const deleteButtons = screen.getAllByRole("button", { name: /удалить/i });
    fireEvent.click(deleteButtons[1]);
    expect(onDelete).toHaveBeenCalledWith("2");
  });
});
