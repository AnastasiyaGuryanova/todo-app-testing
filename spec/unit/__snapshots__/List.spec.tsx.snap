// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`отображение списка задач 1`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;
